src/psyclone/tests/domain/nemo/transformations/nemo_arrayaccess2loop_trans_test.py

```diff
@@ -0,0 +1,690 @@
+        "  enddo\n\n"
+        "end program test\n")
+    reader = FortranReader()
+    psyir = reader.psyir_from_source(input_code)
+    trans = NemoArrayAccess2LoopTrans()
+    outer_loop = psyir.children[0].children[0]
+    assignment = outer_loop.loop_body[0].loop_body[0]
+    index_node = assignment.lhs.children[1]
+    trans.apply(index_node)
+    writer = FortranWriter()
+    result = writer(psyir)
+    assert expected_code in result
+
+
+def test_apply_same_iterator():
+    '''Check that the apply method works as expected if different loop
+    indices use the same iterator in more than one loop index and does
+    not care what name is used as long as it is not the same as the
+    iterator name for this index.
+
+    '''
+    input_code = (
+        "program test\n"
+        "  real :: a(10,10,10)\n"
+        "  integer :: i\n"
+        "  do i=1,5\n"
+        "    a(2*i,i+1,10) = 0.0e0\n"
+        "  end do\n"
+        "end program test")
+    expected_code = (
+        "program test\n"
+        "  real, dimension(10,10,10) :: a\n"
+        "  integer :: i\n  integer :: jk\n\n"
+        "  do i = 1, 5, 1\n"
+        "    do jk = 10, 10, 1\n"
+        "      a(2 * i,i + 1,jk) = 0.0e0\n"
+        "    enddo\n"
+        "  enddo\n\n"
+        "end program test\n")
+    reader = FortranReader()
+    psyir = reader.psyir_from_source(input_code)
+    trans = NemoArrayAccess2LoopTrans()
+    index_node = psyir.children[0].children[0].loop_body[0].lhs.children[2]
+    trans.apply(index_node)
+    writer = FortranWriter()
+    result = writer(psyir)
+    assert expected_code in result
+
+
+def test_apply_index_order():
+    '''Check that the apply method works as expected if loop indices are
+    not used in the same order as array dimensions and does not care
+    what name is used as long as it is not the same as the iterator
+    name for this index.
+
+    '''
+    input_code = (
+        "program test\n"
+        "  real :: a(10,10,10)\n"
+        "  integer :: i,j\n"
+        "  do j=1,10\n"
+        "    do i=1,10\n"
+        "      a(j,i,10) = 0.0e0\n"
+        "    end do\n"
+        "  end do\n"
+        "end program test")
+    expected_code = (
+        "program test\n"
+        "  real, dimension(10,10,10) :: a\n"
+        "  integer :: i\n  integer :: j\n  integer :: jk\n\n"
+        "  do j = 1, 10, 1\n"
+        "    do i = 1, 10, 1\n"
+        "      do jk = 10, 10, 1\n"
+        "        a(j,i,jk) = 0.0e0\n"
+        "      enddo\n"
+        "    enddo\n"
+        "  enddo\n\n"
+        "end program test\n")
+    reader = FortranReader()
+    psyir = reader.psyir_from_source(input_code)
+    trans = NemoArrayAccess2LoopTrans()
+    outer_loop = psyir.children[0].children[0]
+    assignment = outer_loop.loop_body[0].loop_body[0]
+    index_node = assignment.lhs.children[2]
+    trans.apply(index_node)
+    writer = FortranWriter()
+    result = writer(psyir)
+    print(result)
+    assert expected_code in result
```

Comment:
Added.

Reply:
OK, I've tried to extract something generic.

Reply:
Thanks although I actually meant trying to reduce duplication since those tests all have a Fortran fragment that defines a few arrays and then does a doubly-nested loop. However it's a small point so I'm happy to leave it.